From 5bf89a2e9658a28af4f767ee9cc25d666569c414 Mon Sep 17 00:00:00 2001
From: root <root@buildhost>
Date: Thu, 16 Dec 2021 12:44:47 -0600
Subject: [PATCH 2/3] mtd: Add pairing info for NAND

Add pairing info for the NAND so that we may
utilize slc-mode emulation for enhanced reliability and compatibility
with upstream Linux.
---
 drivers/mtd/nand/raw/nand_base.c  | 8 ++++++++
 drivers/mtd/nand/raw/nand_ids.c   | 3 +++
 drivers/mtd/nand/raw/sunxi_nand.c | 1 +
 3 files changed, 12 insertions(+)

diff --git a/drivers/mtd/nand/raw/nand_base.c b/drivers/mtd/nand/raw/nand_base.c
index 3383ebc561..3cf537536c 100644
--- a/drivers/mtd/nand/raw/nand_base.c
+++ b/drivers/mtd/nand/raw/nand_base.c
@@ -4454,6 +4454,10 @@ int nand_detect(struct nand_chip *chip, int *maf_id,
 
 ident_done:
 
+	/* Enable slc-mode to match upstream Linux */
+	chip->options |= NAND_NEED_SCRAMBLING;
+	mtd_set_pairing_scheme(mtd, &dist3_pairing_scheme);
+
 	if (chip->options & NAND_BUSWIDTH_AUTO) {
 		WARN_ON(chip->options & NAND_BUSWIDTH_16);
 		chip->options |= busw;
diff --git a/drivers/mtd/nand/raw/nand_ids.c b/drivers/mtd/nand/raw/nand_ids.c
index d0cfacc69b..a944b168da 100644
--- a/drivers/mtd/nand/raw/nand_ids.c
+++ b/drivers/mtd/nand/raw/nand_ids.c
@@ -63,6 +63,9 @@ struct nand_flash_dev nand_flash_ids[] = {
 		{ .id = {0xad, 0xde, 0x14, 0xa7, 0x42, 0x4a} },
 		  SZ_16K, SZ_8K, SZ_4M, NAND_NEED_SCRAMBLING, 6, 1664,
 		  NAND_ECC_INFO(56, SZ_1K), 1 },
+	{"MT29F32G08CBACAWP",
+		{ .id = {0x2c, 0x68, 0x4, 0x4a, 0xa9, 0x0} },
+		  SZ_4K, SZ_4K, SZ_1M, NAND_NEED_SCRAMBLING, 6, 224, NAND_ECC_INFO(24, SZ_1K) },
 
 	LEGACY_ID_NAND("NAND 4MiB 5V 8-bit",   0x6B, 4, SZ_8K, SP_OPTIONS),
 	LEGACY_ID_NAND("NAND 4MiB 3,3V 8-bit", 0xE3, 4, SZ_8K, SP_OPTIONS),
diff --git a/drivers/mtd/nand/raw/sunxi_nand.c b/drivers/mtd/nand/raw/sunxi_nand.c
index c378f08f68..88fcd6ac31 100644
--- a/drivers/mtd/nand/raw/sunxi_nand.c
+++ b/drivers/mtd/nand/raw/sunxi_nand.c
@@ -1706,6 +1706,7 @@ static int sunxi_nand_chip_init(int node, struct sunxi_nfc *nfc, int devnum)
 	/* Default tR value specified in the ONFI spec (chapter 4.15.1) */
 	nand->chip_delay = 200;
 	nand->controller = &nfc->controller;
+	nand->options |= NAND_NEED_SCRAMBLING;
 	/*
 	 * Set the ECC mode to the default value in case nothing is specified
 	 * in the DT.
-- 
2.25.1
